diff --git a/src/gnome-terminal.schemas.in b/src/gnome-terminal.schemas.in
index 0dc0d92..ee57144 100644
--- a/src/gnome-terminal.schemas.in
+++ b/src/gnome-terminal.schemas.in
@@ -841,6 +841,20 @@
       </locale>
     </schema>
 
+   <schema>
+      <key>/schemas/apps/gnome-terminal/keybindings/copy2</key>
+      <applyto>/apps/gnome-terminal/keybindings/copy2</applyto>
+      <owner>gnome-terminal</owner>
+      <type>string</type>
+      <default>&lt;Ctrl&gt;&lt;Shift&gt;j</default>
+      <locale name="C">
+         <short>Keyboard shortcut to copy text</short>
+         <long>
+	  the same as copy
+	  max support 1024 bytes
+         </long>
+      </locale>
+    </schema>
 
     <schema>
       <key>/schemas/apps/gnome-terminal/keybindings/paste</key>
@@ -860,6 +874,21 @@
       </locale>
     </schema>
 
+
+    <schema>
+      <key>/schemas/apps/gnome-terminal/keybindings/paste2</key>
+      <applyto>/apps/gnome-terminal/keybindings/paste2</applyto>
+      <owner>gnome-terminal</owner>
+      <type>string</type>
+      <default>&lt;Ctrl&gt;&lt;Shift&gt;k</default>
+      <locale name="C">
+         <short>Keyboard shortcut to paste text</short>
+         <long>
+           the same as paste
+         </long>
+      </locale>
+    </schema>
+
     <schema>
       <key>/schemas/apps/gnome-terminal/keybindings/full_screen</key>
       <applyto>/apps/gnome-terminal/keybindings/full_screen</applyto>
diff --git a/src/terminal-accels.c b/src/terminal-accels.c
index c47f558..e02ae62 100644
--- a/src/terminal-accels.c
+++ b/src/terminal-accels.c
@@ -61,7 +61,9 @@
 #define ACCEL_PATH_CLOSE_TAB            ACCEL_PATH_ROOT "FileCloseTab"
 #define ACCEL_PATH_CLOSE_WINDOW         ACCEL_PATH_ROOT "FileCloseWindow"
 #define ACCEL_PATH_COPY                 ACCEL_PATH_ROOT "EditCopy"
+#define ACCEL_PATH_COPY2                ACCEL_PATH_ROOT "EditCopy2"
 #define ACCEL_PATH_PASTE                ACCEL_PATH_ROOT "EditPaste"
+#define ACCEL_PATH_PASTE2               ACCEL_PATH_ROOT "EditPaste2"
 #define ACCEL_PATH_TOGGLE_MENUBAR       ACCEL_PATH_ROOT "ViewMenubar"
 #define ACCEL_PATH_FULL_SCREEN          ACCEL_PATH_ROOT "ViewFullscreen"
 #define ACCEL_PATH_RESET                ACCEL_PATH_ROOT "TerminalReset"
@@ -81,6 +83,7 @@
 #define KEY_CLOSE_TAB           CONF_KEYS_PREFIX "/close_tab"
 #define KEY_CLOSE_WINDOW        CONF_KEYS_PREFIX "/close_window"
 #define KEY_COPY                CONF_KEYS_PREFIX "/copy"
+#define KEY_COPY2               CONF_KEYS_PREFIX "/copy2"
 #define KEY_DETACH_TAB          CONF_KEYS_PREFIX "/detach_tab"
 #define KEY_FULL_SCREEN         CONF_KEYS_PREFIX "/full_screen"
 #define KEY_HELP                CONF_KEYS_PREFIX "/help"
@@ -91,6 +94,7 @@
 #define KEY_NEW_WINDOW          CONF_KEYS_PREFIX "/new_window"
 #define KEY_NEXT_TAB            CONF_KEYS_PREFIX "/next_tab"
 #define KEY_PASTE               CONF_KEYS_PREFIX "/paste"
+#define KEY_PASTE2              CONF_KEYS_PREFIX "/paste2"
 #define KEY_PREV_TAB            CONF_KEYS_PREFIX "/prev_tab"
 #define KEY_RESET_AND_CLEAR     CONF_KEYS_PREFIX "/reset_and_clear"
 #define KEY_RESET               CONF_KEYS_PREFIX "/reset"
@@ -161,6 +165,10 @@ static KeyEntry edit_entries[] =
     KEY_COPY, ACCEL_PATH_COPY, GDK_SHIFT_MASK | GDK_CONTROL_MASK, GDK_KEY (c), NULL, FALSE, TRUE },
   { N_("Paste"),
     KEY_PASTE, ACCEL_PATH_PASTE, GDK_SHIFT_MASK | GDK_CONTROL_MASK, GDK_KEY (v), NULL, FALSE, TRUE },
+  { N_("Copy2"),
+    KEY_COPY2, ACCEL_PATH_COPY2, GDK_SHIFT_MASK | GDK_CONTROL_MASK, GDK_KEY (j), NULL, FALSE, TRUE },
+  { N_("Paste2"),
+    KEY_PASTE2, ACCEL_PATH_PASTE2, GDK_SHIFT_MASK | GDK_CONTROL_MASK, GDK_KEY (k), NULL, FALSE, TRUE },
 };
 
 static KeyEntry view_entries[] =
diff --git a/src/terminal-window.c b/src/terminal-window.c
index 033d7c9..4dabf51 100644
--- a/src/terminal-window.c
+++ b/src/terminal-window.c
@@ -45,6 +45,8 @@
 #include "terminal-util.h"
 #include "terminal-window.h"
 
+static char gCopy2Data[1024];
+
 struct _TerminalWindowPrivate
 {
   GtkActionGroup *action_group;
@@ -176,8 +178,12 @@ static void file_close_tab_callback           (GtkAction *action,
                                                TerminalWindow *window);
 static void edit_copy_callback                (GtkAction *action,
                                                TerminalWindow *window);
+static void edit_copy_callback2               (GtkAction *action,
+                                               TerminalWindow *window);
 static void edit_paste_callback               (GtkAction *action,
                                                TerminalWindow *window);
+static void edit_paste_callback2              (GtkAction *action,
+                                               TerminalWindow *window);
 static void edit_select_all_callback          (GtkAction *action,
                                                TerminalWindow *window);
 static void edit_keybindings_callback         (GtkAction *action,
@@ -976,6 +982,8 @@ terminal_window_update_copy_sensitivity (TerminalScreen *screen,
 
   action = gtk_action_group_get_action (priv->action_group, "EditCopy");
   gtk_action_set_sensitive (action, can_copy);
+  action = gtk_action_group_get_action (priv->action_group, "EditCopy2");
+  gtk_action_set_sensitive (action, can_copy);
 }
 
 static void
@@ -1832,6 +1840,12 @@ terminal_window_init (TerminalWindow *window)
       { "EditPaste", GTK_STOCK_PASTE, NULL, "<shift><control>V",
         NULL,
         G_CALLBACK (edit_paste_callback) },
+      { "EditCopy2", GTK_STOCK_COPY, N_("_Copy2"), "<shift><control>J",
+        NULL,
+        G_CALLBACK (edit_copy_callback2) },
+      { "EditPaste2", GTK_STOCK_PASTE, N_("_Paste2"), "<shift><control>V",
+        NULL,
+        G_CALLBACK (edit_paste_callback2) },
       { "EditPasteURIPaths", GTK_STOCK_PASTE, N_("Paste _Filenames"), "",
         NULL,
         G_CALLBACK (edit_paste_callback) },
@@ -3460,6 +3474,35 @@ typedef struct {
   gboolean uris_as_paths;
 } PasteData;
 
+
+static void
+clipboard_received_cb_text (GtkClipboard *clipboard,
+                             const  char *text,
+                            PasteData *data)
+{
+  if (text != NULL){
+    strncpy(gCopy2Data,text,sizeof(gCopy2Data)-1);
+  }
+}
+
+static void
+edit_copy_callback2 (GtkAction *action,
+                    TerminalWindow *window)
+{
+  TerminalWindowPrivate *priv = window->priv;
+  GtkClipboard *clipboard;
+  const char *name;
+
+  if (!priv->active_screen)
+    return;
+
+  vte_terminal_copy_primary (VTE_TERMINAL (priv->active_screen));
+  clipboard = gtk_widget_get_clipboard (GTK_WIDGET (window), GDK_SELECTION_PRIMARY );
+  gtk_clipboard_request_text (clipboard,
+                                (GtkClipboardURIReceivedFunc) clipboard_received_cb_text,
+                                NULL);
+}
+
 static void
 clipboard_uris_received_cb (GtkClipboard *clipboard,
                             /* const */ char **uris,
@@ -3535,6 +3578,20 @@ edit_paste_callback (GtkAction *action,
                                  data);
 }
 
+
+static void
+edit_paste_callback2 (GtkAction *action,
+                     TerminalWindow *window)
+{
+  TerminalWindowPrivate *priv = window->priv;
+  GtkClipboard *clipboard;
+
+  if (!priv->active_screen)
+    return;
+  if (gCopy2Data != NULL)
+    vte_terminal_feed_child (VTE_TERMINAL (priv->active_screen),gCopy2Data,strlen(gCopy2Data));
+}
+
 static void
 edit_select_all_callback (GtkAction *action,
                           TerminalWindow *window)
diff --git a/src/terminal.xml b/src/terminal.xml
index 3c9cbed..ea24ad6 100644
--- a/src/terminal.xml
+++ b/src/terminal.xml
@@ -15,6 +15,8 @@
     <menu action="Edit">
       <menuitem action="EditCopy" />
       <menuitem action="EditPaste" />
+      <menuitem action="EditCopy2" />
+      <menuitem action="EditPaste2" />
       <menuitem action="EditPasteURIPaths" />
       <separator />
       <menuitem action="EditSelectAll" />
